modules/tensor_mechanics/src/materials/ComputeLagrangianStressCauchy.C

```diff
@@ -0,0 +1,68 @@
+#include "ComputeLagrangianStressCauchy.h"
+
+InputParameters
+ComputeLagrangianStressCauchy::validParams()
+{
+  InputParameters params = ComputeLagrangianStressBase::validParams();
+  return params;
+}
+
+ComputeLagrangianStressCauchy::ComputeLagrangianStressCauchy(const InputParameters & parameters)
+  : ComputeLagrangianStressBase(parameters),
+    _inv_df(getMaterialPropertyByName<RankTwoTensor>("inv_inc_def_grad")),
+    _inv_def_grad(getMaterialPropertyByName<RankTwoTensor>("inv_def_grad")),
+    _detJ(getMaterialPropertyByName<Real>("detJ"))
+{
+}
+
+void
+ComputeLagrangianStressCauchy::computeQpStressUpdate()
+{
+  computeQpCauchyStress();
+  _wrapCauchyStress();
+}
+
+void
+ComputeLagrangianStressCauchy::_wrapCauchyStress()
+{
+  // Actually do the (annoying) wrapping
+  if (_ld)
+  {
+    _pk1_stress[_qp] = _detJ[_qp] * _cauchy_stress[_qp] * _inv_def_grad[_qp].transpose();
+
+    _pk1_jacobian[_qp].zero();
+    for (size_t i = 0; i < 3; i++)
+    {
+      for (size_t J = 0; J < 3; J++)
+      {
+        for (size_t k = 0; k < 3; k++)
+        {
+          for (size_t L = 0; L < 3; L++)
+          {
+            for (size_t m = 0; m < 3; m++)
+            {
+              _pk1_jacobian[_qp](i, J, k, L) +=
+                  _detJ[_qp] * _cauchy_stress[_qp](i, m) *
+                  (_inv_def_grad[_qp](L, k) * _inv_def_grad[_qp](J, m) -
+                   _inv_def_grad[_qp](J, k) * _inv_def_grad[_qp](L, m));
+              for (size_t n = 0; n < 3; n++)
+              {
+                for (size_t t = 0; t < 3; t++)
+                {
+                  _pk1_jacobian[_qp](i, J, k, L) += _detJ[_qp] * _cauchy_jacobian[_qp](i, m, n, t) *
+                                                    _inv_def_grad[_qp](J, m) * _inv_df[_qp](n, k) *
+                                                    _inv_def_grad[_qp](L, t);
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  }
```

Comment:
Let's try to make this (seemingly) less annoying!
```suggestion
  // Actually do the (annoying) wrapping
  if (_ld)
  {
    _pk1_stress[_qp] = _detJ[_qp] * _cauchy_stress[_qp] * _inv_def_grad[_qp].transpose();

    _pk1_jacobian[_qp] = _pk1_stress[_qp].outerProduct(_inv_def_grad[_qp].transpose()) + _pk1_stress[_qp].mixedProductIklJ(_inv_def_grad[_qp]);
    _pk1_jacobian[_qp] += _detJ[_qp] * _cauchy_jacobian[_qp].tripleContraction(_inv_def_grad[_qp].transpose(), _inv_df[_qp], _inv_def_grad[_qp].transpose());
  }
```
You can find the definition of `outerProduct` in `RankTwoTensorImplementation.h`, where you will also find some predefined mixed products. You'll need to write a new `mixedProductIklJ` method. I just made up this `tripleContraction` method, which you could define in `RankFourTensorImplementation.h`.

Reply:
Yep.  You can do the first one with just the mixed product already available, but I added a `tripleContraction` thing.